Cleanup with pylint

svaningelgem opened the pull request:
- Cleanup the pylint things that were ignored. Most notably the cyclic-imports.
- Made all internal references relative. Don't import pysparkling itself anymore from within the library.
- Removed (object) inheritance. It's the only way anyway on Python3.
- `Tokenizer::next` failed on pylint. Likely because pylint confuses it with the internal `next` command? Renamed it to get_next to silence the error... I don't like it, but I prefer to have the switch enabled to capture other use cases.

svenkreiss commented on pysparkling/context.py:
```diff
@@ -194,6 +193,9 @@ def parallelize(self, x, numSlices=None):
         :rtype: RDD
         """
         if numSlices is None or numSlices <= 1:
+            # pylint: disable=import-outside-toplevel, cyclic-import
```
This was importable at the top level before. Can you elaborate why that cannot be done anymore?

svaningelgem replied:
A sample of the error I'm getting:
```
pysparkling\__init__.py:8: in <module>
    from . import exceptions, fileio, streaming
pysparkling\streaming\__init__.py:2: in <module>
    from .context import StreamingContext
pysparkling\streaming\context.py:4: in <module>
    from .dstream import DStream
pysparkling\streaming\dstream.py:5: in <module>
    from ..rdd import EmptyRDD
pysparkling\rdd.py:23: in <module>
    from .context import Context
pysparkling\context.py:17: in <module>
    from .rdd import RDD
E   ImportError: cannot import name 'RDD'
```

The problem is that everything is depending on RDD/Row and Context. I tried playing around a bit with the order of the imports, but I failed to find a good fit.

And both rdd & context have the other imported...
As RDD is only needed 2x in context.py, I opted to make it inline there. (Context appears 165x in rdd.py).

svenkreiss replied:
This was working before. This is a blocker for me. See below in RDD file for a suggestion.

svenkreiss commented on pysparkling/rdd.py:
```diff
@@ -167,9 +167,9 @@ def seqFuncByKey(tc, i):
             return r
 
-        def combFuncByKey(l):
+        def combFuncByKey(lst):
```
To disentangle these PRs and to move forward, I'd suggest not to mix issues and not convert `l` to `lst`. I am against abbreviations in variable names and for some of your replacements `l` is not actually a list so we would have to have a discussion about those names which seems to be out of scope for this PR.

svaningelgem replied:
ok

svenkreiss reviewed (changes requested):
Thanks for all this work. Some comments below.

svaningelgem commented on pysparkling/__version__.py:
```diff
@@ -0,0 +1,3 @@
+from ._version import get_versions
```
Yes indeed. I couldn't fix it in another way because it would recursively call the library and as the first step the version again.

svenkreiss replied:
I have to come back to this. In `context.py` this also happened before: Context is imported from `__init__.py` but inside of `context.py` the `__version__` is imported from `__init__.py`. 

I assume because of the order that `__version__` was the first thing defined, this doesn't lead to a cycle. This should still be possible.

svaningelgem commented on pysparkling/sql/group.py:
```diff
@@ -1,10 +1,9 @@
-from pysparkling.sql.functions import avg, count, lit, max, mean, min, parse, sum
+from .column import Column
+# pylint: disable=redefined-builtin
+from .functions import avg, count, lit, max, mean, min, parse, sum
```
I didn't want to change too much, but I'll fix this.
What do you think about `from . import functions as F`?

svenkreiss replied:
Thanks.

And no, reduce indirections as much as possible.

svenkreiss replied:
Don't use `F`. Avoid indirections.

svaningelgem replied:
```
import functions:
E   ModuleNotFoundError: No module named 'functions'

from . import functions:
*** Works fine

import .functions
E     File "E:\pysparkling\pysparkling\sql\group.py", line 1
E       import .functions
E              ^
E   SyntaxError: invalid syntax
```

So I don't think I have any other way than to call it the way I did? I don't know how to reduce the complexity of the imports here.
Could you provide me some pointers?

Thanks

svenkreiss replied:
There seems to be some misunderstanding. Use `from . import functions`. You say that works fine. Use that.

svenkreiss replied:
The "indirection" to avoid is that someone has to look up what `F` is.

svaningelgem commented on pysparkling/sql/internal_utils/readers/common.py:
```diff
@@ -1,8 +1,8 @@
-from pysparkling.sql.internal_utils.readers.csvreader import CSVReader
```
Same as before, I did not want to cleanup anything here. Just make everything relative.
But I adjust as you said. It's cleaner that way.

svaningelgem commented on pysparkling/sql/internal_utils/readers/utils.py:
```diff
@@ -1,10 +1,10 @@
-from pysparkling.sql.types import (
+from ....fileio import File, TextFile
+from ...casts import get_caster
+from ...types import (
```
I'm a bit on the fence here. Although I would prefer the import as it is now.
isort takes care of doubles if they would appear. Plus, I think the code is more clear when you do not have the `types.` always in front of it...

svaningelgem reviewed (commented):
I went through the comments and wrote my train of thought. Can you check if this changes your mind?

svenkreiss commented on pysparkling/rdd.py:
```diff
@@ -2080,10 +2080,9 @@ def toDF(self, schema=None, sampleRatio=None):
+        # pylint: disable=import-outside-toplevel  # against a cyclic-import
+        from .sql.session import SparkSession
+
         sparkSession = SparkSession._instantiatedSession or SparkSession(Context())
```
I don't know why a new context is created here. The import issues probably come from here. There is a `self.context` that you get without importing `Context`. If you can change those, can you avoid importing Context at the top of the file?

svaningelgem replied:
Thanks for your help! I fixed it as such and now it works fine.

svenkreiss commented on pysparkling/sql/session.py:
```diff
@@ -261,6 +260,9 @@ def prepare(obj):
                 return obj
 
+        # pylint: disable=import-outside-toplevel, cyclic-import
+        from ..rdd import RDD
```
My understanding is that the import structure within the `sql` module is complicated and not great at the moment. However, the core classes outside of the `sql` module should be fine. Do you mind re-checking whether these imports outside of top-level are really necessary for those cases like this one?

svaningelgem replied:
I was able to move a few, but the complete cleanup of these imports would lead me too far. I'm going to move it into an issue (as it very likely will become) to rationalize this later on.

svenkreiss commented on pysparkling/sql/session.py:
```diff
@@ -1,21 +1,20 @@
-from pysparkling.sql.readwriter import DataFrameReader
-from pysparkling.sql.schema_utils import infer_schema_from_list
-from pysparkling.sql.types import (
+from .. import Context, RDD
```
In case circular imports are still blocking, this line might be an issue. Use `from ..context import Context` and `from ..rdd import RDD` instead.

svaningelgem commented:
The only thing which is blocking now is this circular import:
```
scripts\tcpperf_server.py:1:0: R0401: Cyclic import (pysparkling -> pysparkling.context) (cyclic-import)
```

Which is correct. `__init__` is importing Context, and Context is importing the __version__ and accumulators from .

I now changed a few things and the cyclic imports are gone... However they're appearing again when I'm reverting this particular change:
```
sql\session.py: 
from .. import __version__
```

This is way too fickle to my liking. That's the reason why I introduced the __version__ file so you could load the version directly from the file, and don't need to go through loading the whole module.

Additional benefit in my opinion is that the version is in it's own file, you don't need to remove the `get_versions` method anymore. (= cleaner code)

svenkreiss commented:
Thank you for this major effort! LGTM.
